### DISCUSSION TITLE
multi-group support/scancode handling

### DISCUSSION THREAD

**Discussion by minteeaa** [cafali/SnapKey#46] [New Features & Ideas]
# multigroup support 
with the new updates that add multiple key groups, i've extended the functionality to add theoretically infinite key groups with equally infinite amounts of keys within them (with the limitation of how many keys your keyboard has, lol)

[you can test it here on my fork](https://github.com/minteeaa/SnapKey) (build instructions remain the same)

the config can be defined as such:
```cfg
[Group]
key1=...
key2=...

[Group]
key1=...
key2=...
key3=...

[Group]
key1=...
key2=...
key3=...
```
where `[Group]` denotes a new key group to register and key names do not have to be `key1, key2, etc...`; what matters is they begin with `key`, so you can now have names for each such as `key_back, key_forward, key_left, etc...`

the program now handles multiple groups in separate from each other, and will error if there are overlapping keys in the config

# thread blocking (potentially)
while implementing this extension, I took notice that some games like CS:Source and CS1.6 do not respond _at all_ to inputs generated by my fork of SnapKey

it, as I figured, has something to do with Windows' `SendInput`, as the inputs generated are very clearly synthetic to the system; unlike using `XTest` or `uinput` on Linux, where they emulate real keypresses - though the python library `pynput` as far as I know uses the same Windows API to implement the same functionality, so I could be wrong here

i'll keep looking into this, but open to suggestions as I don't have extensive knowledge on specifically this part of the Windows API, being more of a Linux dev for most of my career (and am close to finishing a native Linux implementation of SnapKey ^^)

**Comment by minteeaa** [comment]
# update
after doing some research, I came across [this pynput issue](https://github.com/moses-palmer/pynput/issues/133) that revealed to me SnapKey doesn't handle `scancodes`, essentially a physical representation of a key on an input method that is lower-level than standard `keycode`s: Source Engine, GoldSRC, and many Engines opt to use `scancodes` as a more accurate form of input.

[i've implemented a scancode solution to my fork](https://github.com/minteeaa/SnapKey/commit/3fd9b6ba52d0eb4e3a0de7f45034abe8f497cef8) and after some testing, the following games now respond properly to SnapKey inputs:

- Counter Strike 1.6
- Counter Strike: Source
- Minecraft
- Call of Duty MW3

these are only from my testing, I'm sure many more games listed as incompatible will benefit from the fix.

i'm in the middle of cleaning up the implementation, but bump to @cafali to ask if this is alright to create a PR for?

**Reply by cafali** [reply]
Looks good, go ahead and make a PR when you’re done cleaning things up. 
Thanks a lot for your engagement!
